File: typescript/src/abstractions/sql-dialect.ts
import type { MigrationClient } from '../migration-client.ts';
import type { SqlDialect } from '../sql.ts';

export interface ISqlDialect {
  readonly name: SqlDialect;
  readonly connectionEnvironmentVariables: readonly string[];
  readonly migratesDdl: string;
  readonly migrateLogsDdl: string;
  readonly usesLastInsertRowid: boolean;
  quoteIdent(ident: string): string;
  nowExpr(): string;
  limitClause(n: number): string;
  translatePlaceholders(sql: string): string;
  prerequisiteError(connection: string): Promise<string | null>;
  prepareSetup(connection: string): Promise<void>;
  createClient(connection: string): Promise<MigrationClient>;
}

File: rust/src/abstractions/sql_dialect.rs
pub trait SqlDialect: Send + Sync {
    fn name(&self) -> &'static str;
    fn connection_env_vars(&self) -> &'static [&'static str];
    fn migrates_ddl(&self) -> &'static str;
    fn migrate_logs_ddl(&self) -> &'static str;
    fn use_transaction(&self) -> bool;
    fn select_applied_sql(&self) -> &'static str;
    fn select_last_applied_sql(&self) -> &'static str;
    fn insert_applied_sql(&self) -> &'static str;
    fn delete_applied_sql(&self) -> &'static str;
    fn normalize_connection(&self, connection: &str) -> String;
    fn prerequisite_error(&self, connection: &str) -> Option<String>;
    fn prepare_setup(&self, connection: &str) -> std::io::Result<()>;
}
